### PR TITLE
Prototype of generic view trait

### PR DESCRIPTION
**Pull request by raphlinus**
Use type system hackery to represent a generic bound over traits.

Note: this patch *just* makes the view itself generic, and doesn't address any of the supporting machinery (ViewSequence, AnyView, etc).

### CODE REVIEW

File: src/view/sequence.rs
Comment (xarvic) [inline src/view/sequence.rs:25]: As i said on Zulip, i would like to avoid the explicit mention of `Box` in in `TraitPod `. Xilem#53 did'nt need it so maybe we can avoid it here too? 
But this is probably something we could change later without touching much code, as it only affects the element types.

**Review by xarvic — APPROVED** [review]
With the `TraitPod` trait and the updated version of `ViewSequence` i am in favor of this PR.
I added a comment, but i think we should not block on this.

File: src/view/view.rs
Comment (nicoburns) [inline src/view/view.rs:69]: I'd like to see some better documentation on the generic parameters in this trait (this PR seems to introduce a new one without explaining it, and `A` is never mentioned by name)

File: src/view/view.rs
Comment (nicoburns) [inline src/view/view.rs:102]: Is the idea here that `View` is vello-renderer (=xilem widgets) specific, and `GenericView` could apply to other renderers such as one targetting web, svg, etc? (and that we might therefore get an `SvgView` trait similar to `View` as part of such a backend?)

If so, I'd be tempted to call these `View` and `NativeView` rather than `GenericView` and `View`. But that's probably not important (and could always be changed later). Perhaps just a little note on `View` explaining the difference between it and `GenericView` would be good for now?

File: src/main.rs
Comment (nicoburns) [inline src/main.rs:18]: This seems like a bit of a regression in ergonomics:

- Will end users be expected to write out types like this?
- Will children always be required to be boxed like this?

File: src/widget/core.rs
Comment (nicoburns) [inline src/widget/core.rs:566]: Gah: Pods, Elements, Widgets, Views. I'm still struggling to work out / keep track of what each of these is and how they relate to each other. Some kind of glossary / explainer of the key nouns / concepts in Xilem would be incredibly helpful.

**Review by nicoburns — COMMENTED** [review]
I feel like I've not really been able to provide a substantive review on the key trait changes here, because I feel like I still don't really understand how it all fits together. But I suspect I'm not the only one in that position, so I'm hoping that my questions and requests for documentation will be helpful in their own right.

It's probably a task for after this PR is merged, but I'd *really* like to see a slightly more involved example app built out. Difficult pieces (e.g. text editing) could be avoided, but something with multiple pieces of state and nested views (what do I do once I no longer want to have my entire app in one function? how do I build my own view/widget?)

File: src/widget/core.rs
Comment (jaredoconnell) [reply src/widget/core.rs:566]: I agree that extra documentation for this would be beneficial.
Xilem's architecture won't be immediately intuitive to all first learning it, so the more documentation, the better.

**Review by jaredoconnell — COMMENTED** [review]
While I don't have the insights necessary to foresee all pros and cons of this vs #53, this does look promising.
I don't like that it adds more boilerplate, but the fact that it could lead to entirely different backend implementations could make Xilem very versatile in the future.

I agree with Nico that some extra documentation would be helpful. There are a lot of common terms that have specific meanings in Xilem. Documentation is very helpful in creating a maintainable project.

**Comment by raphlinus** [comment]
I am closing this because I think a macro based approach (#61) is the way forward. In particular, a showstopper is that the trait cannot be made generic over both `Send` and `!Send`, and that is required to support both efficient (multithreaded) native and Web cases.

**Pull request closed without merging**
